pkgs/development/libraries/science/math/nccl/default.nix

```diff
@@ -1,46 +1,40 @@
-{ lib
+{ autoAddOpenGLRunpathHook
 , backendStdenv
+, cuda_cccl
+, cuda_cudart
+, cuda_nvcc
+, cudaFlags
+, cudaVersion
 , fetchFromGitHub
+, lib
 , which
-, cudaPackages ? { }
-, addOpenGLRunpath
 }:
-
-with cudaPackages;
-
-let
-  # Output looks like "-gencode=arch=compute_86,code=sm_86 -gencode=arch=compute_86,code=compute_86"
-  gencode = lib.concatStringsSep " " cudaFlags.gencode;
-in
-backendStdenv.mkDerivation rec {
-  name = "nccl-${version}-cuda-${cudaPackages.cudaMajorVersion}";
+backendStdenv.mkDerivation (finalAttrs: {
+  name = "nccl-${finalAttrs.version}-cuda-${cudaVersion}";
   version = "2.16.5-1";
 
   src = fetchFromGitHub {
     owner = "NVIDIA";
     repo = "nccl";
-    rev = "v${version}";
+    rev = "v${finalAttrs.version}";
     hash = "sha256-JyhhYKSVIqUKIbC1rCJozPT1IrIyRLGrTjdPjJqsYaU=";
   };
 
   outputs = [ "out" "dev" ];
 
   nativeBuildInputs = [
-    which
-    addOpenGLRunpath
+    autoAddOpenGLRunpathHook
+    cuda_cccl #include <nv/target>
+    cuda_cudart #include cuda_runtime.h
```

Comment:
I assume this is what @orivej  refers to: cudart and cccl are headers and shared libs for the target platform, we should try and have them in `buildInputs`

Reply:
Ah I thought `nativeBuildInputs` was for dependencies required at build-time, and `buildInputs` for those required at runtime? If that's not the case, let me know and I'll fix it!

Reply:
Yea, it's a bit more complicated than that and we're being pretty lousy with this only because we don't cross-compile cuda-related packages, we build on the same platform we build _for_

The `nativeBuildInputs` is for the binaries (usually the toolchain) you'd have to execute on your build machine.
The `buildInputs` are for binaries that are to be executed on the target machine, the one you deploy your build artifact at. E.g. if your build machine is x84_64, but you cross-compile for aarch64, you'll link an aarch64 version of `libcudart.so`

In practice, I suspect most of our cuda stuff is not cross-compatible at the moment, not least because of ambiguous string interpolations. But I also just don't know how cross in nixpkgs actually works